mesonpy/__init__.py

```diff
@@ -715,29 +745,13 @@ def __init__(  # noqa: C901
                     self._meson_cross_file.write_text(cross_file_data)
                     self._meson_args['setup'].extend(('--cross-file', os.fspath(self._meson_cross_file)))
 
-        # load config -- PEP 621 support is optional
-        self._config = tomllib.loads(self._source_dir.joinpath('pyproject.toml').read_text())
-        self._pep621 = 'project' in self._config
-        if self.pep621:
-            self._metadata = pyproject_metadata.StandardMetadata.from_pyproject(self._config, self._source_dir)
-        else:
-            print(
-                '{yellow}{bold}! Using Meson to generate the project metadata '
-                '(no `project` section in pyproject.toml){reset}'.format(**_STYLES)
-            )
-            self._metadata = None
-
-        if self._metadata:
-            self._validate_metadata()
+        # load pyproject.toml
+        pyproject = tomllib.loads(self._source_dir.joinpath('pyproject.toml').read_text())
 
-        # load meson args
-        for key in self._get_config_key('args'):
-            self._meson_args[key].extend(self._get_config_key(f'args.{key}'))
-        # XXX: We should validate the user args to make sure they don't conflict with ours.
-
-        self._check_for_unknown_config_keys({
-            'args': _MESON_ARGS_KEYS,
-        })
+        # load meson args from pyproject.toml
+        pyproject_config = _validate_pyproject_config(pyproject)
```

Comment:
Since `pyproject-metadata` is no longer optional, we can use it's API for our own validation.

Reply:
Yup. Working on it. I'm busy fighting mypy...

Reply:
In particular
```
mesonpy/__init__.py:795: error: Argument "version" to "StandardMetadata" has incompatible type "str"; expected "Optional[Version]"  [arg-type]
mesonpy/__init__.py:804: error: Incompatible types in assignment (expression has type "str", variable has type "Optional[Version]")  [assignment]
```
where the second is a preexisting issue. I don't know why mypy in CI does not find it.

Reply:
You need to make it a `packaging.version.Version`.

Reply:
Sure, but I don't understand why it was not flagged before. I didn't change the code at line 804.

Reply:
BTW, which `pyproject-metadata` validation API are you referring too? Looking at the source code I don't see anything related.

Reply:
The `DataFetcher` class. We should replicate `StandardMetadata`, but for our metadata.

Reply:
I'm not sure I understand how it should be used. Does an example exist somewhere? Maybe I'll do that in a separate PR.

Reply:
> `DataFetcher` seems an adapter used by `StandardMetadata` to obtain the metadata field values. I don't see much about validation. Or do you mean to use it to populate the metadata from `meson.build` when there is no metsdata in `pyproject.toml`? It seems not worth the effort to get just the package name and the version.

When you fetch a key via the helper methods, it validates the type, that's what I meant by its validation. Unfortunately, it can't check for unknown keys yet, an initial implementation was provided in #334, but I will finish some of the details I am not happy about and push it `pyproject-metadata`.